romancal/ramp_fitting/ramp_fit_step.py

```diff
@@ -7,76 +7,179 @@
+
+        log.info(f"Using algorithm = {self.algorithm}")
+        log.info(f"Using weighting = {self.weighting}")
+
+        buffsize = ramp_fit.BUFSIZE
+        image_info, integ_info, opt_info, gls_opt_model = ramp_fit.ramp_fit(
+            input_model,
+            buffsize,
+            self.save_opt,
+            readnoise_model.data.value,
+            gain_model.data.value,
+            self.algorithm,
+            self.weighting,
+            max_cores,
+            dqflags.pixel,
+        )
+        readnoise_model.close()
+        gain_model.close()
 
+        # Save the OLS optional fit product, if it exists
+        if opt_info is not None:
+            opt_model = create_optional_results_model(input_model, opt_info)
+            self.save_model(opt_model, "fitopt", output_file=self.opt_name)
+
+        # All pixels saturated, therefore returning an image file with zero data
+        if image_info is None:
+            log.info("All pixels are saturated. Returning a zeroed-out image.")
+
+            # Image info order is: data, dq, var_poisson, var_rnoise, err
+            image_info = (
+                np.zeros(input_model.data.shape[2:], dtype=input_model.data.dtype),
+                input_model.pixeldq
+                | input_model.groupdq[0][0]
+                | dqflags.group["SATURATED"],
+                np.zeros(input_model.err.shape[2:], dtype=input_model.err.dtype),
+                np.zeros(input_model.err.shape[2:], dtype=input_model.err.dtype),
+                np.zeros(input_model.err.shape[2:], dtype=input_model.err.dtype),
+            )
+
+        out_model = create_image_model(input_model, image_info)
+        return out_model
+
+    def ols_cas22(self, input_model, readnoise_model, gain_model):
+        """Peform Optimal Linear Fitting on arbitrarily space resulants
+
+        Parameters
+        ----------
+        input_model : RampModel
+            Model containing ramps.
+
+        readnoise_model : ReadnoiseRefModel
+            Model with the read noise reference information.
+
+        gain_model : GainRefModel
+            Model with the gain reference information.
+
+        Returns
+        -------
+        out_model : ImageModel
+            Model containing a count-rate image.
+        """
+        resultants = input_model.data.value
+        dq = input_model.groupdq
+        read_noise = readnoise_model.data.value
+        gain = gain_model.data.value
+        read_pattern = input_model.meta.exposure.read_pattern
+
+        # add in the gain
+        resultants_adjusted = resultants * gain
+        read_noise_adjusted = read_noise * gain
+
+        # Fit the ramps
+        ramppar, rampvar = ols_cas22_fit.fit_ramps_casertano(
+            resultants_adjusted, dq, read_noise_adjusted, read_pattern=read_pattern
+        )
+
+        # Break out the information and fix units
+        slopes = ramppar[..., 1]
+        var_rnoise = rampvar[..., 0, 1, 1]
+        var_poisson = rampvar[..., 1, 1, 1]
```

Comment:
This reminds me that let's go ahead and abandon my decision to structure these arrays unnecessarily largely.  Maybe parallel the stcal OLS output more closely, or something?  But I feel bad that rampvar has 12 different covariances for every pixel, only one of which is actually populated, which are then dropped.  The RampFitInterpolator populated all of these, of course, and maybe we will down the line for Roman as well, but let's not assume that we'll support that now.

Reply:
Sorry, not clear as to what is desired? It is that `rampvar` needs to have the [...,1 ,1] dimensions removed? Or is there something more needed? Should the `RampData` object be fully supported, or more supported?

Reply:
Presently these lines
https://github.com/spacetelescope/stcal/pull/175/files#diff-9b179303cde3ddf2b749be9d17d7b44e2f93d1207cb0306e8bd955ad41683ff7R233-R238
build 14 detector planes worth of parameters, but only populate four of those planes.  The rationale was that the other ramp fitting code in the simulator computed all 14 planes, but the Casertano+22 paper only discusses the computation of the four done here.  Since we aren't supporting the other ramp fitting code here, there isn't a reason to make these larger shapes.  So the suggestion is to change the interface there to just return the four par / var terms that are computed, and not to package them in these bigger mostly empty objects and then unpackage them.

It would be virtuous to support RampData, though there are a number of quantities there where I don't actually know what they do.  Let's just stick with some simple tuple or dictionary of four items for now unless you see things differently.